refactor(core): bump messageformat parser. Add TS types

thekip opened the pull request:
Bumped version of the upstream messageformat parser to 5.0, updated what's broke and add TS typings wherever possible in touched files. 

Resolves 
https://github.com/lingui/js-lingui/issues/1278
https://github.com/lingui/js-lingui/issues/1303

Should be no changes in logic, but public types slightly changed. They became more accurate, so it's better to bump a version just in case. 

How current config of semver works, should i name a PR with special prefix or commit message or it's doing manually?

andrii-bodnar commented:
@thekip thanks for the contribution!

Probably, I'll ask @Martin005 to review this PR.

> How current config of semver works, should i name a PR with special prefix or commit message or it's doing manually?

It expects that the PR title will be specified according to the following format:

https://github.com/amannn/action-semantic-pull-request#validation

Martin005 commented:
@thekip You made changes only to the `@lingui/core` package.
Could you please also update the `@lingui/cli` and `@lingui/remote-loader`? They currently also use the `messageformat-parser` package.

Also, updating the package may also fix issues #1033 and #1075 🙂

thekip commented:
My bad, didn't check this. 

This commit is bigger, few things about what changed: 

1. I've found a code duplication (3 times) where tokens from `message-parser` were converted to Lingui internal format. 
I removed this code duplication by exporting from `@lingui/core` a `parseMessage` function and reusing it in all subpackages. 

2. I simplified the way how the compiled catalogs produced. This was for 3 purposes: 
   1. to reuse mentioned function
   2. to make code much cleaner and smaller
   3. to make reading of catalogs actually faster than before. 

Before compiled catalogs were produced as js template literal (js code) and complicated AST manipulation was involved
```js
export const messages = {
  messageId: "message value"
}
```
Now i'm just using JSON.stringify / JSON.parse
```js
export const messages = JSON.parse("{
 "messageId": "message value"
}")
```
Why it's better you can read here https://v8.dev/blog/cost-of-javascript-2019#json

Martin005 commented on packages/core/src/i18n.ts:
```diff
@@ -7,23 +7,26 @@ import { EventEmitter } from "./eventEmitter"
 
 export type LocaleData = {
-  plurals?: Function
+  plurals?: (n: number, ordinal?: boolean) => number
```
The return type needs to be `PluralCategory` from the `make-plural` instead of `number`.

Martin005 reviewed (changes requested):
Thank you for the changes, I think that all of what you mentioned makes perfect sense including the JSON.stringify/JSON.parse! 🙂

When I tried your branch as a local package, I immediately noticed one mistake, so **please have a look at my comments and resolve them**.

When I resolved the mistake locally, I also tried all 4 issues #1033, #1075, #1278 and #1303 and I can confirm that all of them are fixed with this new version of `@messageformat/parser` package 🥳

thekip commented:
Fixed

thekip commented:
I also thinking on how to hide exposed but treated as internal APIs, for example that `formatMessage` is supposed to be internal api, but because we use it in other packages it should be exported. 

There should be either special naming convention, or may be just JSdoc would be enough

thekip commented:
I see tests are failing, i will have a look

thekip commented:
I've ported some changes from this PR https://github.com/lingui/js-lingui/pull/1328

> I deleted pure option from createCompiledCatalog because it's not used anywhere in the project. This API considered internal, so we should be on the safe side.
It's probably was created for remote-loader but code has been changed since then and now this option is obsolete.

And also i've ported tests, because previous ones didn't test anything (there was a logical issue)

Martin005 reviewed (approved):
Great, thank you for the changes!
